Switch rules new canonical Dart compiler and align APIs with Google's internal rules

jelbourn opened the pull request:
This PR:
* Switches the default compiler from [libsass](https://github.com/sass/libsass) (via [sassc](https://github.com/sass/sassc)) to [dart-sass](https://github.com/sass/dart-sass) (via the [JavaScript-compiled distribution on npm](https://www.npmjs.com/package/sass)). The Dart compiler is now the canonical implementation and is actively maintained by the Sass owners.
* Updates the rule APIs to match the API of Google's internal Sass rules so that these rules can then _replace_ the existing Sass rules inside Google. This is largely adding additional attributes to `sass_binary`, excepting the breaking changes listed below.

### New features
Sass binary now supports three new attributes: `include_paths`, `output_dir`, and `output_name`. See the README for more information on these attributes.

### Breaking changes
* Additional setup in a project's `WORKSPACE` file is now required; see the README for more information.
* The default output filename for `sass_binary` is now based on the `src` filename instead of the name of the rule. For example, if your input `src` file is `material-button.scss`, the default output filename will be `material-button.css`. The name can be customized via the `output_name` attribute.
* While the Dart compiler should match the libsass compiler's behavior, it is possible some minor differences in CSS output may occur.


(I'm taking over this change for @alexeagle; still including his original commits, which in turn include code from @nex3)

alexeagle commented on README.md:
```diff
@@ -128,253 +131,51 @@ Target //hello_world:hello_world up-to-date:
-
-<a name="reference-sass_library"></a>
-## sass_library
+#### Implicit output targets
```
We can generate the documentation with skydoc from the inline PyDoc comments - but I suppose it's too hard to host the resulting documentation and point users there in this change?
We should remove this copy of the documentation since it will get stale quickly.

jelbourn replied:
I intend for this to be temporarily until we're hosting the skydoc

alexeagle commented on sass/package.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "devDependencies": {
+    "sass": "1.3.2"
```
we should have doc somewhere for maintainers to know how to cut releases

jelbourn replied:
Let's do that in a follow-up PR, mainly because I don't know how to do that myself

nex3 commented on sass/sass.bzl:
```diff
@@ -11,219 +11,170 @@
+   As of 2018, this is the most commonly used compiler.
+   It requires your Bazel setup has a working C++ compilation toolchain.
+
+   NOTE: future releases of rules_sass may remove this option, if it
```
Unless we have someone who's volunteering to actively maintain LibSass support, I think we should bite the bullet and drop it immediately, especially if support is going to require extra work to make command-line flags line up properly.

jelbourn replied:
For now it looks like `sourcemap` is the only one supported that's problematic. How about leaving it for now, and dropping it if we run into more mismatches?

nex3 replied:
I'm worried about a couple things. First, saying we "support LibSass" when we actually only support a [two-and-a-half-year-old beta release](https://github.com/sass/libsass/releases?after=3.3.0) is pretty misleading, especially when we *will* be supporting up-to-date versions of Dart Sass. Second, I'm worried that if people start relying on these implementations being swappable and that property breaks later on (which it may very well), we'll have unexpected and painful breakage for downstream users.

jelbourn replied:
Removed all references to libsass/sassc in docstrings

nex3 commented on sass/sass.bzl:
```diff
@@ -11,219 +11,170 @@
+
+2. Dart Sass runs the new compiler implementation natively in the Dart
+   VM. This option requires a change to the Dart Bazel rules which is
+   not yet available as of May 2018.
```
It would be nice to link to a bug for this. @natebosch, do you know where we're tracking Bazel Dart support?

natebosch replied:
The rules are at https://github.com/dart-lang/rules_dart 

We do not have active plans or a timeline for revisiting external bazel support.

nex3 replied:
The readme says the rules are "in active development"... it might be a good idea to update that to give a better idea of their current status.

@jelbourn I guess the best thing to do here is link to https://github.com/dart-lang/rules_dart and say "when this is actively maintained".

natebosch replied:
Good idea. Updated

jelbourn replied:
Done

nex3 commented on sass/sass.bzl:
```diff
@@ -11,219 +11,170 @@
+  args = ctx.actions.args()
+
+  # Flags (see https://github.com/sass/node-sass#options)
+  # Note, the command line is compatible with Dart sass.
```
There's some overlap between Dart Sass's and Node Sass's command-line options, but they aren't identical and we don't intend to retain compatibility on principle.

jelbourn replied:
Is the JS output version of the Dart compiler API-equivalent with node-sass? I had remembered you mentioning that it was supposed to be a drop-in replacement. 

Is there a better place to point for the Dart compiler flags?

nex3 replied:
> Is the JS output version of the Dart compiler API-equivalent with node-sass? I had remembered you mentioning that it was supposed to be a drop-in replacement.

It's JS-API-equivalent, but not CLI-equivalent. We expect most automation to go through the JS API, and for the CLI to be generally human-facing; Bazel is unusual in that regard.

> Is there a better place to point for the Dart compiler flags?

Unfortunately there isn't (again because we tend to expect the CLI to be human-facing). I guess [the source file where they're defined](https://github.com/sass/dart-sass/blob/master/lib/src/executable/options.dart)?

jelbourn replied:
Removed the mention of API equivalence and pointed to the source file.

jelbourn commented on README.md:
```diff
@@ -2,37 +2,42 @@
     name = "io_bazel_rules_sass",
-    remote = "https://github.com/bazelbuild/rules_sass.git",
-    tag = "0.0.3",
+    # Make sure to check for the latest version when you install
```
@alexeagle where is that information?

alexeagle replied:
The releases tab on GitHub is where I look.
Ideally we should just update the Readme as part of releases, I think maybe the go rules do that?

jelbourn commented on README.md:
```diff
@@ -128,253 +131,51 @@ Target //hello_world:hello_world up-to-date:
+| Label            | Description                                                                  |
+|------------------|------------------------------------------------------------------------------|
+| **name**.css     | The generated CSS output                                                     |
+| **name**.css.map | The [source map][sm] that can be used to debug the Sass source in-browser    |
```
Cool, I had no idea you could do that. Done.

jelbourn commented on README.md:
````diff
@@ -128,253 +131,51 @@ Target //hello_world:hello_world up-to-date:
-```python
-sass_binary(name, src, deps=[], output_style="compressed")
+```py
+sass_binary(name, src, deps=[], output_style="compressed", include_paths=[], output_dir=".",output_name=<src_filename.css>)
````
Space added; I'm okay leaving it long because the attributes are broken down below.

nex3 replied:
SGTM

jelbourn commented:
I tried consuming these rules from Angular Material and made a couple last fixes:
* Moved the skydoc rule to a separate package so that downstream apps don't have to add skydoc to their workspace
* Fixed CI by disabling skydoc generation on Windows
* Namespaced the dependencies in sass_repositories function to be scoped to the sass workspace
